pallets/asset-tx-payment/src/lib.rs

```diff
@@ -0,0 +1,256 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Asset Transaction Payment Module
+//!
+//! This module provides the basic logic needed to pay the absolute minimum amount needed for a
+//! transaction to be included via the assets (other than the main token of the chain).
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_std::prelude::*;
+use codec::{Encode, Decode, EncodeLike};
+use frame_support::{DefaultNoBound, decl_module, decl_storage, dispatch::DispatchResult, traits::{Get, IsType}, weights::{
+		Weight, DispatchInfo, PostDispatchInfo, GetDispatchInfo, Pays, WeightToFeePolynomial,
+		WeightToFeeCoefficient, DispatchClass,
+	}};
+use sp_runtime::{
+	FixedU128, FixedPointNumber, FixedPointOperand, Perquintill, RuntimeDebug,
+	transaction_validity::{
+		InvalidTransaction, TransactionPriority, ValidTransaction, TransactionValidityError, TransactionValidity,
+	},
+	traits::{
+		Saturating, SignedExtension, SaturatedConversion, Convert, Dispatchable,
+		DispatchInfoOf, PostDispatchInfoOf, Zero, One,
+	},
+};
+use pallet_balances::NegativeImbalance;
+use pallet_transaction_payment::OnChargeTransaction;
+use frame_support::traits::tokens::{fungibles::{Balanced, Inspect, CreditOf}, WithdrawConsequence};
+
+#[cfg(test)]
+mod tests;
+
+mod payment;
+pub use payment::*;
+
+pub(crate) type BalanceOf<T> = <<T as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<T>>::Balance;
+pub(crate) type AssetBalanceOf<T> = <<T as Config>::Fungibles as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+pub(crate) type ChargeAssetBalanceOf<T> = <<T as pallet::Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::Balance;
+pub(crate) type AssetIdOf<T> = <<T as Config>::Fungibles as Inspect<<T as frame_system::Config>::AccountId>>::AssetId;
+pub(crate) type ChargeAssetIdOf<T> = <<T as pallet::Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::AssetId;
+pub(crate) type LiquidityInfoOf<T> = <<T as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<T>>::LiquidityInfo;
+pub(crate) type ChargeAssetLiquidityOf<T> = <<T as pallet::Config>::OnChargeAssetTransaction as OnChargeAssetTransaction<T>>::LiquidityInfo;
+
+#[derive(Encode, Decode, DefaultNoBound)]
+pub enum InitialPayment<T: Config> {
+	Nothing,
+	Native(LiquidityInfoOf<T>),
+	Asset(CreditOf<T::AccountId, T::Fungibles>),
+}
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	use frame_support::{
+		dispatch::DispatchResultWithPostInfo,
+		pallet_prelude::*,
+		inherent::Vec,
+		traits::{
+			Currency, ReservableCurrency, EnsureOrigin, ExistenceRequirement::KeepAlive,
+		},
+		PalletId,
+	};
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_transaction_payment::Config + pallet_balances::Config {
```

Comment:
just starting to read this, but do we need to hard-depend on pallets here?

Reply:
for `pallet_transaction_payment` see below comment.
`pallet_balances` might be a remnant from an earlier implementation, will try to remove and see what happens :-D